gobblin-runtime/src/main/java/org/apache/gobblin/runtime/mapreduce/MRJobLauncher.java

```diff
@@ -170,6 +181,12 @@ public MRJobLauncher(Properties jobProps, Configuration conf, SharedResourcesBro
       throws Exception {
     super(jobProps, metadataTags);
 
+    this.fsm = new FiniteStateMachine.Builder<MRJobLauncherState>()
```

Comment:
This FSM instance (and the associated states) looks generic enough that it can be moved out of this class. Can we have a FSM factory class that builds out specific FSM types and hands it back to the caller?